In [1]:
%%time

import pandas as pd

train = pd.read_csv('/kaggle/input/playground-series-s4e7/train.csv')
test = pd.read_csv('/kaggle/input/playground-series-s4e7/test.csv')
sample_submission = pd.read_csv('/kaggle/input/playground-series-s4e7/sample_submission.csv')

CPU times: user 23.7 s, sys: 4.06 s, total: 27.8 s
Wall time: 36.7 s


In [2]:
%%time

train.head(30)

CPU times: user 176 µs, sys: 31 µs, total: 207 µs
Wall time: 195 µs


id  Gender  Age  Driving_License  Region_Code  Previously_Insured  \
0    0    Male   21                1         35.0                   0   
1    1    Male   43                1         28.0                   0   
2    2  Female   25                1         14.0                   1   
3    3  Female   35                1          1.0                   0   
4    4  Female   36                1         15.0                   1   
5    5  Female   31                1         47.0                   1   
6    6    Male   23                1         45.0                   1   
7    7  Female   47                1          8.0                   0   
8    8  Female   26                1         28.0                   1   
9    9  Female   66                1         11.0                   0   
10  10    Male   22                1          3.0                   1   
11  11  Female   25                1         10.0                   0   
12  12    Male   36                1         28.0                   0   
13  13    Male   27                1         28.0                   1   
14  14  Female   24                1         12.0                   1   
15  15    Male   79                1          8.0                   0   
16  16    Male   34                1         37.0                   0   
17  17  Female   20                1         50.0                   1   
18  18  Female   25                1         30.0                   1   
19  19  Female   27                1         30.0                   1   
20  20  Female   34                1         50.0                   1   
21  21    Male   34                1         46.0                   1   
22  22  Female   43                1         29.0                   1   
23  23  Female   24                1          8.0                   0   
24  24  Female   42                1         11.0                   0   
25  25  Female   23                1         15.0                   0   
26  26  Female   43                1         35.0                   1   
27  27  Female   23                1         11.0                   1   
28  28    Male   40                1         35.0                   0   
29  29    Male   40                1         28.0                   0   

   Vehicle_Age Vehicle_Damage  Annual_Premium  Policy_Sales_Channel  Vintage  \
0     1-2 Year            Yes         65101.0                 124.0      187   
1    > 2 Years            Yes         58911.0                  26.0      288   
2     < 1 Year             No         38043.0                 152.0      254   
3     1-2 Year            Yes          2630.0                 156.0       76   
4     1-2 Year             No         31951.0                 152.0      294   
5     < 1 Year             No         28150.0                 152.0      197   
6     < 1 Year             No         27128.0                 152.0      190   
7     1-2 Year            Yes         40659.0                  26.0      262   
8     < 1 Year             No         31639.0                 152.0       36   
9     1-2 Year            Yes          2630.0                  26.0      125   
10    < 1 Year             No         27996.0                 152.0      215   
11    < 1 Year            Yes          2630.0                 152.0       30   
12    1-2 Year            Yes         38104.0                  26.0      203   
13    < 1 Year             No         43746.0                 152.0      259   
14    < 1 Year             No         23692.0                 152.0      245   
15   > 2 Years            Yes         66887.0                 124.0      150   
16    1-2 Year            Yes         30226.0                  26.0      240   
17    < 1 Year             No         36389.0                 160.0      242   
18    < 1 Year             No         25769.0                 152.0      117   
19    < 1 Year             No         40797.0                 152.0      151   
20    1-2 Year             No          2630.0               

In [3]:
%%time

test.head(30)

CPU times: user 117 µs, sys: 20 µs, total: 137 µs
Wall time: 129 µs


id  Gender  Age  Driving_License  Region_Code  Previously_Insured  \
0   11504798  Female   20                1         47.0                   0   
1   11504799    Male   47                1         28.0                   0   
2   11504800    Male   47                1         43.0                   0   
3   11504801  Female   22                1         47.0                   1   
4   11504802    Male   51                1         19.0                   0   
5   11504803    Male   22                1         30.0                   1   
6   11504804  Female   23                1         20.0                   0   
7   11504805  Female   21                1         37.0                   1   
8   11504806    Male   72                1         43.0                   1   
9   11504807    Male   33                1         28.0                   1   
10  11504808    Male   67                1         48.0                   0   
11  11504809    Male   34                1         29.0                   0   
12  11504810  Female   25                1         28.0                   1   
13  11504811  Female   24                1         47.0                   1   
14  11504812    Male   45                1         11.0                   1   
15  11504813  Female   65                1         28.0                   0   
16  11504814    Male   40                1         28.0                   1   
17  11504815  Female   24                1         21.0                   1   
18  11504816    Male   35                1         10.0                   1   
19  11504817    Male   40                1          0.0                   0   
20  11504818    Male   47                1         36.0                   0   
21  11504819  Female   26                1         28.0                   0   
22  11504820    Male   41                1         28.0                   0   
23  11504821    Male   66                1         28.0                   0   
24  11504822    Male   24                1         41.0                   1   
25  11504823    Male   31                1         28.0                   0   
26  11504824  Female   39                1         28.0                   1   
27  11504825  Female   24                1         45.0                   1   
28  11504826  Female   80                1         28.0                   0   
29  11504827    Male   52                1         28.0                   1   

   Vehicle_Age Vehicle_Damage  Annual_Premium  Policy_Sales_Channel  Vintage  
0     < 1 Year             No          2630.0                 160.0      228  
1     1-2 Year            Yes         37483.0                 124.0      123  
2     1-2 Year            Yes          2630.0                  26.0      271  
3     < 1 Year             No         24502.0                 152.0      115  
4     1-2 Year             No         34115.0                 124.0      148  
5     < 1 Year             No         25715.0                 152.0      245  
6     < 1 Year            Yes          2630.0                 152.0       11  
7     < 1 Year            Yes          2630.0                 160.0      171  
8     1-2 Year             No         38495.0                 124.0       97  
9     < 1 Year             No         46937.0                 152.0       72  
10    1-2 Year            Yes          2630.0                  29.0      225  
11    < 1 Year             No          2630.0                   1.0      102  
12    < 1 Year             No         29574.0                 152.0       36  
13    < 1 Year             No         36601.0                 152.0      292  
14    1-2 Year             No          2630.0                 122.0      250  
15    1-2 Year             No          2630.0                  26.0      152  
16    1-2 Year             No         58712.0                  26.0      278  
17    < 1 Year             No         28963.0                 152.0       55  
18    1-2 Year            Yes          2630.0              

In [4]:
%%time

# check missing values

missing_values = train.isnull().sum()
missing_values

CPU times: user 3.15 s, sys: 31.7 ms, total: 3.18 s
Wall time: 3.18 s


id                      0
Gender                  0
Age                     0
Driving_License         0
Region_Code             0
Previously_Insured      0
Vehicle_Age             0
Vehicle_Damage          0
Annual_Premium          0
Policy_Sales_Channel    0
Vintage                 0
Response                0
dtype: int64

# Use h2o automl to train model

* I used h2o automl to handle training large dataset . 

In [5]:
%%time

import h2o
print(h2o.__version__)
from h2o.automl import H2OAutoML

h2o.init(max_mem_size='16G')

3.46.0.4
Checking whether there is an H2O instance running at http://localhost:54321..... not found.
Attempting to start a local H2O server...
  Java Version: openjdk version "11.0.24" 2024-07-16; OpenJDK Runtime Environment (build 11.0.24+8-post-Ubuntu-1ubuntu322.04); OpenJDK 64-Bit Server VM (build 11.0.24+8-post-Ubuntu-1ubuntu322.04, mixed mode, sharing)
  Starting server from /opt/conda/lib/python3.10/site-packages/h2o/backend/bin/h2o.jar
  Ice root: /tmp/tmpsn7m4bpx
  JVM stdout: /tmp/tmpsn7m4bpx/h2o_unknownUser_started_from_python.out
  JVM stderr: /tmp/tmpsn7m4bpx/h2o_unknownUser_started_from_python.err
  Server is running at http://127.0.0.1:54321
Connecting to H2O server at http://127.0.0.1:54321 ... successful.


--------------------------  ----------------------------------
H2O_cluster_uptime:         02 secs
H2O_cluster_timezone:       Etc/UTC
H2O_data_parsing_timezone:  UTC
H2O_cluster_version:        3.46.0.4
H2O_cluster_version_age:    1 month and 25 days
H2O_cluster_name:           H2O_from_python_unknownUser_xyh3df
H2O_cluster_total_nodes:    1
H2O_cluster_free_memory:    16 Gb
H2O_cluster_total_cores:    4
H2O_cluster_allowed_cores:  4
H2O_cluster_status:         locked, healthy
H2O_connection_url:         http://127.0.0.1:54321
H2O_connection_proxy:       {"http": null, "https": null}
H2O_internal_security:      False
Python_version:             3.10.14 final
--------------------------  ----------------------------------

CPU times: user 314 ms, sys: 79.3 ms, total: 394 ms
Wall time: 7.71 s


In [6]:
%%time

train = h2o.import_file("/kaggle/input/playground-series-s4e7/train.csv")
test = h2o.import_file("/kaggle/input/playground-series-s4e7/test.csv")
sample_submission = h2o.import_file("/kaggle/input/playground-series-s4e7/sample_submission.csv")

Parse progress: |████████████████████████████████████████████████████████████████| (done) 100%
Parse progress: |████████████████████████████████████████████████████████████████| (done) 100%
Parse progress: |████████████████████████████████████████████████████████████████| (done) 100%
CPU times: user 584 ms, sys: 76.1 ms, total: 660 ms
Wall time: 25.8 s


In [7]:
%%time

x = train.columns[1:]
y = 'Response'

CPU times: user 18 µs, sys: 0 ns, total: 18 µs
Wall time: 22.4 µs


In [8]:
%%time

# For binary classification, response should be a factor
train[y] = train[y].asfactor()

CPU times: user 239 µs, sys: 28 µs, total: 267 µs
Wall time: 270 µs


In [9]:
%%time

aml = H2OAutoML(max_models=200, seed=1000, max_runtime_secs=1000)
aml.train(x=x, y=y, training_frame=train)

AutoML progress: |███████████████████████████████████████████████████████████████| (done) 100%
CPU times: user 6.84 s, sys: 675 ms, total: 7.52 s
Wall time: 16min 58s


Model Details
=============
H2OGradientBoostingEstimator : Gradient Boosting Machine
Model Key: GBM_1_AutoML_1_20240904_133057


Model Summary: 
    number_of_trees    number_of_internal_trees    model_size_in_bytes    min_depth    max_depth    mean_depth    min_leaves    max_leaves    mean_leaves
--  -----------------  --------------------------  ---------------------  -----------  -----------  ------------  ------------  ------------  -------------
    75                 75                          2.99006e+06            0            15           10.0667       1             6969          3173.79

ModelMetricsBinomial: gbm
** Reported on train data. **

MSE: 0.0812963545001875
RMSE: 0.2851251558529824
LogLoss: 0.24867423214366913
Mean Per-Class Error: 0.22347833731764127
AUC: 0.8849346271919885
AUCPR: 0.4596376811415074
Gini: 0.7698692543839769

Confusion Matrix (Act/Pred) for max f1 @ threshold = 0.2574963709529308
       0            1            Error    Rate
-----  -----------  -----------  -------  ---------------------
0      5.92879e+06  1.13237e+06  0.1604   (1132370.0/7061162.0)
1      283740       706313       0.2866   (283740.0/990053.0)
Total  6.21253e+06  1.83868e+06  0.1759   (1416110.0/8051215.0)

Maximum Metrics: Maximum metrics at their respective thresholds
metric                       threshold    value        idx
---------------------------  -----------  -----------  -----
max f1                       0.257496     0.499384     206
max f2                       0.152106     0.661862     283
max f0point5                 0.351828     0.459149     135
max accuracy                 0.454005     0.882287     74
max precision                0.792977     0.942857     0
max recall                   0.00113261   1            399
max specificity              0.792977     1            0
max absolute_mcc             0.203929     0.440995     247
max min_per_class_accuracy   0.221509     0.798175     234
max mean_per_class_accuracy  0.141799     0.817692     290
max tns                      0.792977     7.06116e+06  0
max fns                      0.792977     990020       0
max fps                      0.00113261   7.06116e+06  399
max tps                      0.00113261   990053       399
max tnr                      0.792977     1            0
max fnr                      0.792977     0.999967     0
max fpr                      0.00113261   1            399
max tpr                      0.00113261   1            399

Gains/Lift Table: Avg response rate: 12.30 %, avg score: 12.31 %
group    cumulative_data_fraction    lower_threshold    lift         cumulative_lift    response_rate    score       cumulative_response_rate    cumulative_score    capture_rate    cumulative_capture_rate    gain      cumulative_gain    kolmogorov_smirnov
-------  --------------------------  -----------------  -----------  -----------------  ---------------  ----------  --------------------------  ------------------  --------------  -------------------------  --------  -----------------  --------------------
1        0.0100001                   0.526025           5.42177      5.42177            0.666712         0.571785    0.666712                    0.571785            0.0542183       0.0542183                  442.177   442.177            0.0504181
2        0.0200001                   0.484513           4.63724      5.02951            0.570238         0.503233    0.618475                    0.53751             0.0463723       0.100591                   363.724   402.951            0.0918902
3        0.0300001                   0.458211           4.27049      4.7765             0.525139         0.470713    0.587363                    0.515244            0.0427048       0.143295                   327.049   377.65             0.129181
4        0.04                        0.438073           4.02494      4.58861            0.494945         0.44767     0.564259                    0.498351            0.0402494       0.183545                   302.49

* i used 1000 random seed for big data in training . 

42 = small data

1000 = big data

In [10]:
%%time

# View the AutoML Leaderboard
lb = aml.leaderboard
lb.head(rows=lb.nrows)

CPU times: user 3.32 ms, sys: 2.06 ms, total: 5.38 ms
Wall time: 8.59 ms


model_id                                 auc    logloss     aucpr    mean_per_class_error      rmse        mse
----------------------------------  --------  ---------  --------  ----------------------  --------  ---------
GBM_1_AutoML_1_20240904_133057      0.87827    0.253158  0.438343                0.234562  0.28733   0.0825583
XGBoost_1_AutoML_1_20240904_133057  0.877858   0.252787  0.435414                0.229146  0.287399  0.0825982
XGBoost_2_AutoML_1_20240904_133057  0.876286   0.254061  0.42965                 0.227508  0.288151  0.0830312
GLM_1_AutoML_1_20240904_133057      0.841239   0.272436  0.329876                0.244493  0.298633  0.0891819
[4 rows x 7 columns]

* Metric roc-auc score is 0.87827 . The model making a good performance .

In [11]:
%%time

preds = aml.predict(test)

gbm prediction progress: |███████████████████████████████████████████████████████| (done) 100%
CPU times: user 297 ms, sys: 37.6 ms, total: 335 ms
Wall time: 1min 17s


In [12]:
%%time

# Extracting the probability for class 1 (positive response)
preds_prob = preds['p1']

CPU times: user 136 µs, sys: 20 µs, total: 156 µs
Wall time: 160 µs


In [13]:
%%time

# Convert H2OFrame to pandas DataFrame
preds_prob_df = preds_prob.as_data_frame()

/opt/conda/lib/python3.10/site-packages/h2o/frame.py:1981: H2ODependencyWarning: Converting H2O frame to pandas dataframe using single-thread.  For faster conversion using multi-thread, install polars and pyarrow and use it as pandas_df = h2o_df.as_data_frame(use_multi_thread=True)

  warnings.warn("Converting H2O frame to pandas dataframe using single-thread.  For faster conversion using"


CPU times: user 2.24 s, sys: 902 ms, total: 3.14 s
Wall time: 5.4 s


In [14]:
%%time

# Load test data
test_df = h2o.as_list(test)  # Converts H2O test data to pandas DataFrame
submission = pd.DataFrame()
submission['id'] = test_df['id']  # Assuming 'id' column exists in test dataset
submission['Response'] = preds_prob_df  # Adding the predicted probabilities

/opt/conda/lib/python3.10/site-packages/h2o/frame.py:1981: H2ODependencyWarning: Converting H2O frame to pandas dataframe using single-thread.  For faster conversion using multi-thread, install polars and pyarrow and use it as pandas_df = h2o_df.as_data_frame(use_multi_thread=True)

  warnings.warn("Converting H2O frame to pandas dataframe using single-thread.  For faster conversion using"


CPU times: user 9.82 s, sys: 3.63 s, total: 13.4 s
Wall time: 20.6 s


In [15]:
%%time

# Save the submission to a CSV file
submission.to_csv('submission.csv', index=False)

CPU times: user 20.6 s, sys: 227 ms, total: 20.8 s
Wall time: 20.8 s


In [16]:
submission.head()

id  Response
0  11504798  0.017081
1  11504799  0.424550
2  11504800  0.245157
3  11504801  0.001143
4  11504802  0.029030

#  Calculate the percentage of responses

In [21]:
%%time

# Convert the H2OFrame to a pandas DataFrame
predictions_df = preds.as_data_frame()

/opt/conda/lib/python3.10/site-packages/h2o/frame.py:1981: H2ODependencyWarning: Converting H2O frame to pandas dataframe using single-thread.  For faster conversion using multi-thread, install polars and pyarrow and use it as pandas_df = h2o_df.as_data_frame(use_multi_thread=True)

  warnings.warn("Converting H2O frame to pandas dataframe using single-thread.  For faster conversion using"


CPU times: user 4.81 s, sys: 1.91 s, total: 6.72 s
Wall time: 11.2 s


In [22]:
%%time

# Convert probabilities to labels (0 or 1) based on a threshold of 0.5
predictions_df['Label'] = predictions_df['p1'].apply(lambda x: 1 if x >= 0.5 else 0)

CPU times: user 5.27 s, sys: 255 ms, total: 5.53 s
Wall time: 5.53 s


In [24]:
%%time

# Count the number of each prediction (0: No Response, 1: Response)
response_counts = predictions_df['Label'].value_counts()

# Total number of predictions
total = len(predictions_df)

total

CPU times: user 62.4 ms, sys: 20 µs, total: 62.4 ms
Wall time: 61.9 ms


7669866

* Model made predictions for 7,669,866 customers 

In [25]:
%%time

# Respond
percentage_respond = (response_counts[1] / total) * 100  # Percentage of '1' (Respond)

print(f"Percentage of Customers Who Responded: {percentage_respond:.2f}%")

Percentage of Customers Who Responded: 1.54%
CPU times: user 277 µs, sys: 41 µs, total: 318 µs
Wall time: 319 µs


* Estimated 118,107 customers are predicted to accept the insurance offer 

In [26]:
%%time


# Not Respond

percentage_not_respond = (response_counts[0] / total) * 100  # Percentage of '0' (Not Respond)

print(f"Percentage of Customers Who Did Not Respond: {percentage_not_respond:.2f}%")

Percentage of Customers Who Did Not Respond: 98.46%
CPU times: user 143 µs, sys: 0 ns, total: 143 µs
Wall time: 141 µs


* Estimated 7,551,759 customers are predicted to not accept the insurance offer 